src/ngraph/ngraph_emitter.cc

```diff
@@ -1285,6 +1286,119 @@ void Emitter::CreateLayerOps() {
+      axes.insert(1);
+    } else if (get_default(node, "preserve_shape", false)) {
+      axes.insert(in_shape.size() - 1);
+    } else {
+      for (size_t i = 1; i < in_shape.size(); ++i) {
+        axes.insert(i);
+      }
+    }
+    return std::make_shared<ngraph::op::Softmax>(input, axes);
+  };
+}
+
+void Emitter::CreateLossOps() {
+  // These functions are in place to provide a mechanism for manually specifying
+  // backpropgation methods for Loss functions. We do this because MXNet
+  // provides a number of options that only effect the output of backprop, not
+  // forward prop, and are difficult or impossible to integrate into
+  // nGraph's autodiff functionality
+  loss_op_backward_funcs_["SoftmaxOutput"] = [this](
+      const NodePtr& node, const NgraphNodePtr& adjoint) {
+    float grad_scale = get_default(node, "grad_scale", 1.0f);
+    float ignore_label = get_default(node, "ignore_label", -1.0f);
+    float smooth_alpha = get_default(node, "smooth_alpha", 0.0f);
+
+    bool use_ignore = get_default(node, "use_ignore", false);
+    bool out_grad = get_default(node, "out_grad", false);
+
+    std::string norm = get_default(node, "normalization", std::string("null"));
+
+    auto softmax = op_map_[node];
+    auto label = op_map_[node->inputs_[1]];
+    bool ignore = false;
+    NgraphNodePtr mask;
+
+    if (label->get_shape() != softmax->get_shape()) {
+      if (use_ignore) {
+        ignore = true;
+        mask = cast_result(
+            std::make_shared<ngraph::op::NotEqual>(
+                label, makeConstant(node, std::to_string(ignore_label))),
+            getType(node->dtype_));
+      }
+      size_t axis = op_map_[node->inputs_[0]]->get_shape().size() - 1;
+      if (get_default(node, "multi_output", false)) {
+        axis = 1;
+      }
+      label = std::make_shared<ngraph::op::OneHot>(label, softmax->get_shape(),
+                                                   axis);
+    }
+
+    if (smooth_alpha != 0.0f) {
+      int num_classes = 1;
+      auto shape = softmax->get_shape();
+      if (get_default(node, "multi_output", false)) {
+        num_classes = shape[1];
+      } else if (get_default(node, "preserve_shape", false)) {
+        num_classes = shape.back();
+      } else {
+        for (size_t i = 1; i < shape.size(); ++i) {
+          num_classes *= shape[i];
+        }
+      }
+      auto one = makeConstant(node, "1");
+      auto smooth_const = makeConstant(node, std::to_string(smooth_alpha));
+      auto subtractions = label * smooth_const;
+      auto additions = (one - label) * smooth_const /
+                       makeConstant(node, std::to_string(num_classes - 1));
+      label = label - subtractions + additions;
+    }
+
+    auto gradient = softmax - label;
+
+    if (ignore) {
+      gradient =
+          ngraph::builder::make_with_numpy_broadcast<ngraph::op::Multiply>(
+              gradient, mask);
+    }
+    if (grad_scale != 1.0f) {
```

Comment:
Comparing two floating-point numbers for equality is a notoriously risky proposition.  Are you sure it's the best approach here?

Reply:
Yeah, it's the same hardcoded 1.0f default above, if they've given me anything that isn't the default, I want to run this branch.